### PR TITLE
 Improve loading efficiency of Timelog on Dashboards page

### PR DESCRIPTION
**Pull request by aaronleechan**
**Install** node-cache : https://www.npmjs.com/package/node-cache

**Add** Cache in taskController.js & userProfilerController.js

**Add** nodeCache.js file under the utilities file.

cache.removeCache user id: deleteUserProfile, putUserProfile, changeUserStatus

cache.removeCache task id: importTask, postTask

timeEntryController.js **update** path **getInfringmentEmailBody** 

userProfileController.js & userProfileJobs.js 

**update** import to const because cannot run npm run serve 

let me know if I required to change back.

### CODE REVIEW

File: src/controllers/taskController.js [hunk 1]
```diff
@@ -562,7 +563,8 @@ const taskController = function (Task) {
       res.status(400).send({ error: 'Task Name, Active status, Task Number are mandatory fields' });
       return;
     } */
-
+    /* Remove Cache If Data Exist */
+    cache.removeCache(req.params.id);
```
Comment (chrisweilacker) [inline src/controllers/taskController.js:567]: just to make absolutely sure it is absolutely unique may want to do ("task-" + req.params.id)

File: src/controllers/taskController.js [hunk 2]
```diff
@@ -673,6 +675,8 @@ const taskController = function (Task) {
       return;
     }
 
+    /* Remove Cache If Data Exist */
+    cache.removeCache(req.params.id);
```
Comment (chrisweilacker) [inline src/controllers/taskController.js:679]: Same thing just to make absolutely sure it is absolutely unique may want to do ("task-" + req.params.id)

File: src/controllers/taskController.js [hunk 3]
```diff
@@ -1068,8 +1072,17 @@ const taskController = function (Task) {
...
+    if (cache.getCache(taskId)) {
+      const getData = cache.getCache(taskId);
+      res.status(200).send(getData);
+      return;
+    }
+
     Task.findById(taskId, '-__v  -createdDatetime -modifiedDatetime')
-      .then(results => res.status(200).send(results))
+      .then((results) => {
+        cache.setCache(taskId, results);
```
Comment (chrisweilacker) [inline src/controllers/taskController.js:1083]: ("task-" + taskId)

File: src/controllers/taskController.js [hunk 4]
```diff
@@ -1068,8 +1072,17 @@ const taskController = function (Task) {
   const getTaskById = function (req, res) {
     const taskId = req.params.id;
 
+    if (cache.getCache(taskId)) {
```
Comment (chrisweilacker) [inline src/controllers/taskController.js:1075]: getCache("task-"+taskId)

File: src/controllers/userProfileController.js [hunk 5]
```diff
@@ -183,6 +183,7 @@ const userProfileController = function (UserProfile) {
       res.status(403).send('You are not authorized to update this user');
       return;
     }
+    cache.removeCache(userid);
```
Comment (chrisweilacker) [inline src/controllers/userProfileController.js:186]: same thing with user: "user-" + userid etc

**Review by chrisweilacker — APPROVED** [review]
Please change the following then you can merge after testing.  But also make sure the master is not out of date.

**Comment by aaronleechan** [comment]
@chrisweilacker  would you please take a look. I already updated it.

File: src/utilities/nodeCache.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
+
+
+const NodeCache = require('node-cache');
+const logger = require('../startup/logger');
+
+const cacheStore = new NodeCache({ stdTTL: 100, checkperiod: 120 });
```
Comment (chrisweilacker) [inline src/utilities/nodeCache.js:6]: Maybe have stdTTL as something larger than 100 seconds maybe set that to 300(every 5 mins) and checkperiod to 600(every 10 mins)

**Review by chrisweilacker — APPROVED** [review]

**Comment by chrisweilacker** [comment]
Looks good I will do the change and merge it. Please have jae put it into dev and test it.  Make sure it improves the speed and that azll the edit/delete functions work in dev.

**Comment by one-community** [comment]
Will get this done by tomorrow afternoon at the latest. 

What do you mean by this though: " azll the edit/delete functions work in dev."

**Pull request merged**